VersionControlUI/src/DiffManager.cpp

```diff
@@ -0,0 +1,288 @@
+		if (change.oldNode_)
+			oldNodes.insert(change.oldNode_, change.changeType_);
+		if (change.newNode_)
+			newNodes.insert(change.newNode_, change.changeType_);
+	}
+
+	QSet<Model::Node*> oldNodesMarkedForRemoval = findAllNodesWithDirectParentPresent(oldNodes);
+	QSet<Model::Node*> newNodesMarkedForRemoval = findAllNodesWithDirectParentPresent(newNodes);
+
+	auto it = container->begin();
+	while (it != container->end())
+	{
+		if (oldNodesMarkedForRemoval.contains(it->oldNode_) || newNodesMarkedForRemoval.contains(it->newNode_))
+			it = container->erase(it);
+		else
+			it++;
+	}
+}
+
+QSet<Model::Node*> DiffManager::findAllNodesWithDirectParentPresent(QHash<Model::Node*,
+																						  FilePersistence::ChangeType>& nodes)
+{
+	QSet<Model::Node*> nodesMarkedForRemoval;
+	for (auto node : nodes.keys())
+		// check that changeType also matches
+		if (nodes.contains(node->parent()) && nodes[node] == nodes[node->parent()])
+				nodesMarkedForRemoval.insert(node);
+
+	return nodesMarkedForRemoval;
+}
+
+// TODO find good way to return and use Node instead of Id
+bool DiffManager::findChangedNode(Model::TreeManager* treeManager, Model::NodeIdType id, Model::NodeIdType& resultId)
+{
+	if (auto node = const_cast<Model::Node*>(treeManager->nodeIdMap().node(id)))
+	{
+
+		Model::Node* changedNode = nullptr;
+		Model::SymbolMatcher symbolMatcher{symbol_};
+
+		if (auto ancestorNode = node->firstAncestorOfType(symbolMatcher))
+		{
+			changedNode = ancestorNode;
+
+			resultId = treeManager->nodeIdMap().id(changedNode);
+			return true;
+		}
+	}
+	return false;
+}
+
+void DiffManager::createOverlaysForChanges(Visualization::ViewItem* diffViewItem,
+														 QList<ChangeWithNodes> changesWithNodes)
+{
+	for (auto change : changesWithNodes)
+	{
+		QString highlightOverlayStyle;
+		QString highlightOverlayName;
+
+		switch (change.changeType_)
+		{
+			case FilePersistence::ChangeType::Deletion:
+				highlightOverlayName = "delete_highlights";
+				highlightOverlayStyle = "delete_no_bg_solid_outline";
+				break;
+			case FilePersistence::ChangeType::Insertion:
+				highlightOverlayName = "insert_highlights";
+				highlightOverlayStyle = "insert_no_bg_solid_outline";
+				break;
+			case FilePersistence::ChangeType::Move:
+				highlightOverlayName = "move_highlights";
+				highlightOverlayStyle = "move_no_bg_solid_outline";
+
+				// add arrow for the moved node
+				diffViewItem->addArrow(const_cast<Model::Node*>(change.oldNode_),
+											  const_cast<Model::Node*>(change.newNode_), QString{"move_arrows"});
+				break;
+			case FilePersistence::ChangeType::Stationary:
+				highlightOverlayName = "modify_highlights";
+				highlightOverlayStyle = "modify_no_bg_solid_outline";
+				break;
+			case FilePersistence::ChangeType::Unclassified:
+				// TODO check what to do in this case
+				throw VersionControlUIException{"Unclassified change found"};
+		}
+
+		Visualization::VisualizationManager::instance().mainScene()->addPostEventAction(
```

Comment:
Hmm, you're adding postevents in each loop iteration, that's not a good idea.

Instead of this, can't you only add one event, that directly calls the `createOverlayrsForChanges` method? Keep in mind that by the time you call this method, the class will probably be destructed, so you'll have to make the method static and pass in all necessary info.